### PR TITLE
feat: support tracing scalars

### PR DESCRIPTION
**Pull request by avik-pal**
realized I need this to correctly compile conditionals with scalars assignments in the branches

```julia
using Reactant

x = (3, [3.14])

function f(x)
    return x[1] * x[2]
end

x_ra = Reactant.to_rarray(x; track_numbers=(Number,))

f2 = @compile f(x_ra)

f2(Reactant.to_rarray((5, [3.14]); track_numbers=(Number,)))
# 1-element ConcreteRArray{Float64, 1} with indices 1:
#  15.700000000000001

x_ra2 = Reactant.to_rarray(x)

f3 = @compile f(x_ra2)

f3(Reactant.to_rarray((5, [3.14])))
# 1-element ConcreteRArray{Float64, 1} with indices 1:
#  9.42
```

### CODE REVIEW

File: src/Tracing.jl [hunk 1]
```diff
@@ -400,8 +415,13 @@ function make_tracer(
 end
 
 function make_tracer(
-    seen, @nospecialize(prev::RT), @nospecialize(path), mode; kwargs...
-) where {RT<:AbstractFloat}
+    seen, @nospecialize(prev::RT), @nospecialize(path), mode; track_numbers=(), kwargs...
+) where {RT<:Number}
+    if mode == ArrayToConcrete
+        length(track_numbers) == 0 && return prev
+        should_convert = any(Base.Fix1(<:, RT), track_numbers)
```
Comment (wsmoses) [inline src/Tracing.jl:422]: is there a reason to not just have this be a boolean flag?

Comment (avik-pal) [reply src/Tracing.jl:422]: In the meeting notes we had this as:

```
x2 = to_rarray(x; track_intscalars=false, track_floatscalars=false) # if scalars (ints) were converted
```

Comment (wsmoses) [reply src/Tracing.jl:422]: ah yeah fair enough

**Review by wsmoses — APPROVED** [review]

File: src/ConcreteRArray.jl [hunk 2]
```diff
@@ -4,11 +4,34 @@ end
...
+    data::T; client=XLA.default_backend[], idx=XLA.default_device_idx[]
+) where {T<:Number}
+    Base.depwarn(
+        "ConcreteRArray(data::Number) is deprecated, use ConcreteRNumber(data) instead",
+        :ConcreteRArray,
+    )
+    return ConcreteRArray(fill(data); client, idx)
+end
+
+const ConcreteRNumberType{T} = Union{ConcreteRArray{T,0},ConcreteRNumber{T}}
```
Comment (mofeing) [inline src/ConcreteRArray.jl:34]: small aesthetic suggestion: how about `ConcreteRScalar`?

Comment (avik-pal) [reply src/ConcreteRArray.jl:34]: Didn't we decide to not do TracedRScalar to match julia's naming of Number?

Comment (mofeing) [reply src/ConcreteRArray.jl:34]: exactly, that's why you have `ConcreteRNumber` and I'm fine with that.

by `*RScalar` I'm referring to this `Union{*RArray{T,0}, *RNumber{T}}`

**Pull request merged**